# Bike_암기 복습_20180912

In [1]:
print("테스트")

테스트


# 데이터, 패키지 로딩

In [2]:
import pandas as pd
import numpy as np

train = pd.read_csv("data/bike/train.csv", parse_dates = ["datetime"])
test = pd.read_csv("data/bike/test.csv", parse_dates = ["datetime"])

print(train.shape, test.shape)
train.head()

(10886, 12) (6493, 9)


datetime  season  holiday  workingday  weather  temp   atemp  \
0 2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1 2011-01-01 01:00:00       1        0           0        1  9.02  13.635   
2 2011-01-01 02:00:00       1        0           0        1  9.02  13.635   
3 2011-01-01 03:00:00       1        0           0        1  9.84  14.395   
4 2011-01-01 04:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40  
2        80        0.0       5          27     32  
3        75        0.0       3          10     13  
4        75        0.0       0           1      1

# Feature Engineering

In [3]:
print(train.info(), "\n", "\n" , test.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 12 columns):
datetime      10886 non-null datetime64[ns]
season        10886 non-null int64
holiday       10886 non-null int64
workingday    10886 non-null int64
weather       10886 non-null int64
temp          10886 non-null float64
atemp         10886 non-null float64
humidity      10886 non-null int64
windspeed     10886 non-null float64
casual        10886 non-null int64
registered    10886 non-null int64
count         10886 non-null int64
dtypes: datetime64[ns](1), float64(3), int64(8)
memory usage: 1020.6 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6493 entries, 0 to 6492
Data columns (total 9 columns):
datetime      6493 non-null datetime64[ns]
season        6493 non-null int64
holiday       6493 non-null int64
workingday    6493 non-null int64
weather       6493 non-null int64
temp          6493 non-null float64
atemp         6493 non-null float64
humidity      6493 non-null

## (1) NaN 처리

- 없음

## (2) Datetime 처리

In [4]:
train["year"] = train["datetime"].dt.year
train["month"] = train["datetime"].dt.month
train["day"] = train["datetime"].dt.day
train["hour"] = train["datetime"].dt.hour
train["dayofweek"] = train["datetime"].dt.dayofweek

train["minute"] = train["datetime"].dt.minute
train["second"] = train["datetime"].dt.second

In [5]:
train.columns
train.head()

datetime  season  holiday  workingday  weather  temp   atemp  \
0 2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1 2011-01-01 01:00:00       1        0           0        1  9.02  13.635   
2 2011-01-01 02:00:00       1        0           0        1  9.02  13.635   
3 2011-01-01 03:00:00       1        0           0        1  9.84  14.395   
4 2011-01-01 04:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  casual  registered  count  year  month  day  hour  \
0        81        0.0       3          13     16  2011      1    1     0   
1        80        0.0       8          32     40  2011      1    1     1   
2        80        0.0       5          27     32  2011      1    1     2   
3        75        0.0       3          10     13  2011      1    1     3   
4        75        0.0       0           1      1  2011      1    1     4   

   dayofweek  minute  second  
0          5       0       0  
1          5       0       0  
2          5       0       0  
3          5       0       0  
4          5       0       0

In [6]:
train[['year', 'month', 'day', 'hour', 'minute', 'second', 'dayofweek']].describe()
# minute, second 정보는 없으므로 drop시킨다.

train.drop(['minute', 'second'], axis = 1, inplace = True)
train.head()

datetime  season  holiday  workingday  weather  temp   atemp  \
0 2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1 2011-01-01 01:00:00       1        0           0        1  9.02  13.635   
2 2011-01-01 02:00:00       1        0           0        1  9.02  13.635   
3 2011-01-01 03:00:00       1        0           0        1  9.84  14.395   
4 2011-01-01 04:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  casual  registered  count  year  month  day  hour  \
0        81        0.0       3          13     16  2011      1    1     0   
1        80        0.0       8          32     40  2011      1    1     1   
2        80        0.0       5          27     32  2011      1    1     2   
3        75        0.0       3          10     13  2011      1    1     3   
4        75        0.0       0           1      1  2011      1    1     4   

   dayofweek  
0          5  
1          5  
2          5  
3          5  
4          5

In [7]:
test["year"] = test["datetime"].dt.year
test["month"] = test["datetime"].dt.month
test["day"] = test["datetime"].dt.day
test["hour"] = test["datetime"].dt.hour
test["dayofweek"] = test["datetime"].dt.dayofweek

test.head()

datetime  season  holiday  workingday  weather   temp   atemp  \
0 2011-01-20 00:00:00       1        0           1        1  10.66  11.365   
1 2011-01-20 01:00:00       1        0           1        1  10.66  13.635   
2 2011-01-20 02:00:00       1        0           1        1  10.66  13.635   
3 2011-01-20 03:00:00       1        0           1        1  10.66  12.880   
4 2011-01-20 04:00:00       1        0           1        1  10.66  12.880   

   humidity  windspeed  year  month  day  hour  dayofweek  
0        56    26.0027  2011      1   20     0          3  
1        56     0.0000  2011      1   20     1          3  
2        56     0.0000  2011      1   20     2          3  
3        56    11.0014  2011      1   20     3          3  
4        56    11.0014  2011      1   20     4          3

# EDA, Visualization

## 시각화 패키지 로딩

In [8]:
import seaborn as sns
import matplotlib as mpl
import matplotlib.pyplot as plt

%matplotlib inline

# RandomForestRegressor

In [9]:
from sklearn.ensemble import RandomForestRegressor

model_rf = RandomForestRegressor()

model_rf

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

# RMSE

In [10]:
from sklearn.metrics import make_scorer

def rmse(predict, actual):
    predict = np.array(predict)
    actual = np.array(actual)
    
    distance = predict - actual
    
    square_distance = distance **2
    mean_square_distance = square_distance.mean()
    score = np.sqrt(mean_square_distance)
    
    return score

rmse_score = make_scorer(rmse)
rmse_score

make_scorer(rmse)

# RF - GridSearch

In [11]:
train.columns

Index(['datetime', 'season', 'holiday', 'workingday', 'weather', 'temp',
       'atemp', 'humidity', 'windspeed', 'casual', 'registered', 'count',
       'year', 'month', 'day', 'hour', 'dayofweek'],
      dtype='object')

# Kaggle Evaluation Metrics에 맞게 변형

In [12]:
train["log_count"] = np.log(train["count"] + 1)

train[["count", "log_count"]].head()

count  log_count
0     16   2.833213
1     40   3.713572
2     32   3.496508
3     13   2.639057
4      1   0.693147

In [13]:
feature_names = ['season', 'holiday', 'workingday', 'weather', 'temp',
       'atemp', 'humidity', 'windspeed',
       'year', 'month', 'day', 'hour', 'dayofweek']

label_name = 'log_count'

X_train = train[feature_names]
X_test = test[feature_names]
y_train = train[label_name]

from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import cross_val_score

n_estimators = 3
max_depth_list = [10, 20]
max_features_list = [0.3, 0.5, 0.9]

hyperparameters_list = []

for max_depth in max_depth_list:
    for max_features in max_features_list:
        model = RandomForestRegressor(n_estimators=n_estimators,
                                      max_depth=max_depth,
                                      max_features=max_features,
                                      n_jobs=-1,
                                      random_state=37)
        score = cross_val_score(model, X_train, y_train, cv=20, scoring=rmse_score).mean()
        hyperparameters = {
            'score': score,
            'n_estimators': n_estimators,
            'max_depth': max_depth,
            'max_features': max_features,
        }
        hyperparameters_list.append(hyperparameters)
        print(f"n_estimators = {n_estimators}, max_depth = {max_depth:2}, max_features = {max_features:.6f}, Score = {score:.5f}")

hyperparameters_list = pd.DataFrame.from_dict(hyperparameters_list)

hyperparameters_list = hyperparameters_list.sort_values(by="score")

print(hyperparameters_list.shape)

hyperparameters_list.head()

In [14]:
from sklearn.model_selection import cross_val_score

In [43]:
#소요시간 측정
import datetime
now = datetime.datetime.today()
print("start:", now)

# 178초. 2분 32초 (Local), 2분 53초 (Google Colab)
n_estimators_list = [100]
max_depth_list = [20, 40]
max_features_list = [0.1, 0.3, 0.5, 0.7, 0.9]

#Grid 변수 - Google Colab. 28분소요
#n_estimators_list = [100, 300, 500]
#max_depth_list = [20, 40, 60, 80, 100]
#max_features_list = [0.1, 0.3, 0.5, 0.7, 0.9]

hyperparameters_list = []

for n_estimators in n_estimators_list:
    for max_depth in max_depth_list:
        for max_features in max_features_list:
            model_rf = RandomForestRegressor(n_estimators=n_estimators,
                                         max_depth = max_depth,
                                         max_features = max_features,
                                         n_jobs = -1,
                                         random_state = 79)
            score = cross_val_score(model_rf, X_train, y_train, cv = 10, scoring = rmse_score).mean()
            
            hyperparameters = {
                'score' : score,
                'n_estimators' : n_estimators,
                'max_depth' : max_depth,
                'max_features' : max_features,
            }
            # score 다음에 쉽표로 구분하지 않아서 생기는 문제들
            
            hyperparameters_list.append(hyperparameters)
            
            print(f"n_estimators = {n_estimators}, max_depth = {max_depth:2}, max_features = {max_features:.2f}, Score = {score:.4f}") # 마지막 따옴표 쳐야


now2 = datetime.datetime.today()
print("end:", now2)
gap = now2- now

m = gap.seconds //60
s = gap.seconds %60

print("Gap Time:", m,"분", s, "초")

start: 2018-09-13 12:00:11.684462
n_estimators = 100, max_depth = 20, max_features = 0.10, Score = 0.6955
n_estimators = 100, max_depth = 20, max_features = 0.30, Score = 0.5055
n_estimators = 100, max_depth = 20, max_features = 0.50, Score = 0.4048
n_estimators = 100, max_depth = 20, max_features = 0.70, Score = 0.3877
n_estimators = 100, max_depth = 20, max_features = 0.90, Score = 0.3830
n_estimators = 100, max_depth = 40, max_features = 0.10, Score = 0.6915
n_estimators = 100, max_depth = 40, max_features = 0.30, Score = 0.5038
n_estimators = 100, max_depth = 40, max_features = 0.50, Score = 0.4058
n_estimators = 100, max_depth = 40, max_features = 0.70, Score = 0.3886
n_estimators = 100, max_depth = 40, max_features = 0.90, Score = 0.3853
end: 2018-09-13 12:02:44.091180
Gap Time: 2 분 32 초


In [166]:
hyperparameters_list

[{'score': 0.38296375490152207,
  'n_estimators': 100,
  'max_depth': 20,
  'max_features': 0.9},
 {'score': 0.6955473574772217,
  'n_estimators': 100,
  'max_depth': 20,
  'max_features': 1}]

In [167]:
hyperparameters_list = pd.DataFrame.from_dict(hyperparameters_list)
hyperparameters_list.head()

max_depth  max_features  n_estimators     score
0         20           0.9           100  0.382964
1         20           1.0           100  0.695547

In [168]:
hyperparameters_list = hyperparameters_list.sort_values(by = "score")
hyperparameters_list.head()

# 0.34까지 나왔음 (DSSchool)
# 0.38까지 나옴 

max_depth  max_features  n_estimators     score
0         20           0.9           100  0.382964
1         20           1.0           100  0.695547

# Coarse Search + Finer Search

In [175]:
n_estimators = 100
num_epoch = 100

coarse_hyperparameters_list = []

for epoch in range(num_epoch):
    max_depth = np.random.randint(low = 2, high = 100)
    max_features = np.random.uniform(low = 0.1, high = 1.0)
    model = RandomForestRegressor(n_estimators = n_estimators,
                                 max_depth = max_depth,
                                 max_features = max_features,
                                 n_jobs = -1,
                                 random_state = 79)
    score = cross_val_score(model, X_train, y_train, cv = 10, scoring = rmse_score).mean()
    
    hyperparameters = {
        'epoch' : epoch,
        'score' : score,
        'n_estimators' : n_estimators,
        'max_depth' : max_depth,
        'max_features' : max_features,
    }
    
    coarse_hyperparameters_list.append(hyperparameters)
    
    print(f"{epoch:2} n_estimators = {n_estimators}, max_depth = {max_depth:2}, max_features = {max_features:.2f}, Score = {score:.4f}")

#    print(f"max_depth: {max_depth:2}, max_features: {max_features:.2f}")

coarse_hyperparameters_list = pd.DataFrame.from_dict(coarse_hyperparameters_list)
coarse_hyperparameters_list = coarse_hyperparameters_list.sort_values(by = "score")
print(coarse_hyperparameters_list.shape)
coarse_hyperparameters_list.head()
# 0.384695

# 44번 돌렸는데 최고 수치는 
# 40 n_estimators = 100, max_depth = 19, max_features = 0.85, Score = 0.3845

 0 n_estimators = 100, max_depth = 81, max_features = 0.79, Score = 0.3847
 1 n_estimators = 100, max_depth = 36, max_features = 0.93, Score = 0.3849
 2 n_estimators = 100, max_depth = 83, max_features = 0.82, Score = 0.3847
 3 n_estimators = 100, max_depth = 61, max_features = 0.96, Score = 0.3849
 4 n_estimators = 100, max_depth =  6, max_features = 0.85, Score = 0.5644
 5 n_estimators = 100, max_depth = 66, max_features = 0.70, Score = 0.3886
 6 n_estimators = 100, max_depth = 99, max_features = 0.81, Score = 0.3847
 7 n_estimators = 100, max_depth = 62, max_features = 0.51, Score = 0.4058
 8 n_estimators = 100, max_depth = 22, max_features = 0.49, Score = 0.4095
 9 n_estimators = 100, max_depth = 41, max_features = 0.24, Score = 0.5038
10 n_estimators = 100, max_depth = 70, max_features = 0.84, Score = 0.3847
11 n_estimators = 100, max_depth = 78, max_features = 0.56, Score = 0.3959
12 n_estimators = 100, max_depth = 83, max_features = 0.49, Score = 0.4058
13 n_estimators = 100, ma

epoch  max_depth  max_features  n_estimators     score
74     74         25      0.887101           100  0.382423
66     66         24      0.901121           100  0.383554
40     40         19      0.847148           100  0.384504
97     97         35      0.829557           100  0.384655
33     33         99      0.788319           100  0.384695

(10, 5)


epoch  max_depth  max_features  n_estimators     score
2      2         71      0.784852           100  0.384695
9      9         75      0.845158           100  0.384695
4      4         64      0.501282           100  0.405799
6      6         10      0.763733           100  0.419860
0      0         12      0.530929           100  0.423197

# Finer Search

In [176]:
n_estimators = 100
num_epoch = 100

coarse_hyperparameters_list = []

for epoch in range(num_epoch):
    max_depth = np.random.randint(low = 19, high = 25)
    max_features = np.random.uniform(low = 0.84, high = 0.91)
    model = RandomForestRegressor(n_estimators = n_estimators,
                                 max_depth = max_depth,
                                 max_features = max_features,
                                 n_jobs = -1,
                                 random_state = 79)
    score = cross_val_score(model, X_train, y_train, cv = 10, scoring = rmse_score).mean()
    
    hyperparameters = {
        'epoch' : epoch,
        'score' : score,
        'n_estimators' : n_estimators,
        'max_depth' : max_depth,
        'max_features' : max_features,
    }
    
    coarse_hyperparameters_list.append(hyperparameters)
    
    print(f"{epoch:2} n_estimators = {n_estimators}, max_depth = {max_depth:2}, max_features = {max_features:.6f}, Score = {score:.5f}")

#    print(f"max_depth: {max_depth:2}, max_features: {max_features:.2f}")

coarse_hyperparameters_list = pd.DataFrame.from_dict(coarse_hyperparameters_list)
coarse_hyperparameters_list = coarse_hyperparameters_list.sort_values(by = "score")
print(coarse_hyperparameters_list.shape)
coarse_hyperparameters_list.head()

 0 n_estimators = 100, max_depth = 21, max_features = 0.902095, Score = 0.38659
 1 n_estimators = 100, max_depth = 19, max_features = 0.903428, Score = 0.38450
 2 n_estimators = 100, max_depth = 23, max_features = 0.847874, Score = 0.38236
 3 n_estimators = 100, max_depth = 20, max_features = 0.895761, Score = 0.38296
 4 n_estimators = 100, max_depth = 19, max_features = 0.887628, Score = 0.38450
 5 n_estimators = 100, max_depth = 21, max_features = 0.907698, Score = 0.38659
 6 n_estimators = 100, max_depth = 22, max_features = 0.860717, Score = 0.38369
 7 n_estimators = 100, max_depth = 23, max_features = 0.874618, Score = 0.38236
 8 n_estimators = 100, max_depth = 22, max_features = 0.864812, Score = 0.38369
 9 n_estimators = 100, max_depth = 22, max_features = 0.889001, Score = 0.38369
10 n_estimators = 100, max_depth = 23, max_features = 0.889757, Score = 0.38236
11 n_estimators = 100, max_depth = 24, max_features = 0.898002, Score = 0.38355
12 n_estimators = 100, max_depth = 23, m

epoch  max_depth  max_features  n_estimators     score
43     43         24      0.840814           100  0.382282
25     25         23      0.885454           100  0.382361
81     81         23      0.859490           100  0.382361
2       2         23      0.847874           100  0.382361
31     31         23      0.908665           100  0.382361

### GridSearch 
- n_estimators = 100, max_depth = 20, max_features = 0.90, Score = 0.3829...
### FinerSearch
- n_estimators = 100, max_depth = 63, max_features = 0.798523, Score = 0.38470


epoch	max_depth	max_features	n_estimators	score
43	43	24	0.840814	100	0.382282
25	25	23	0.885454	100	0.382361
81	81	23	0.859490	100	0.382361

# Fit 

In [ ]:
model_rf.fit(X_train, y_train)
predictions = model_rf.predict(X_test)

predictions

In [47]:
submission = pd.read_csv("data/bike/samplesubmission.csv")

print(submission.shape)
submission.head()

(6493, 2)


datetime  count
0  2011-01-20 00:00:00      0
1  2011-01-20 01:00:00      0
2  2011-01-20 02:00:00      0
3  2011-01-20 03:00:00      0
4  2011-01-20 04:00:00      0

In [48]:
submission["count"] = predictions

print(submission.shape)
submission.head()

(6493, 2)


datetime  count
0  2011-01-20 00:00:00   12.3
1  2011-01-20 01:00:00    8.0
2  2011-01-20 02:00:00    5.3
3  2011-01-20 03:00:00    3.1
4  2011-01-20 04:00:00    2.1

In [49]:
submission.to_csv("data/bike/submission_bike_20180912.csv", index = False)